src/Adapter/Product/ProductDuplicator.php

```diff
@@ -0,0 +1,535 @@
+     * @throws CoreException
+     */
+    public function duplicate(ProductId $productId): ProductId
+    {
+        //@todo: don't forget to add hooks from PrestaShopBundle\Controller\Admin\ProductController L1063 when new controller is implemented
+        //@todo: add database transaction. blocked by PR #20518
+        $product = $this->productRepository->get($productId);
+        $oldProductId = $productId->getValue();
+        $newProduct = $this->duplicateProduct($product);
+        $newProductId = (int) $newProduct->id;
+
+        $this->duplicateRelations($newProductId, $oldProductId);
+
+        if ($product->hasAttributes()) {
+            $this->updateDefaultAttribute($newProductId, $oldProductId);
+        }
+
+        $this->hookDispatcher->dispatchWithParameters(
+            'actionProductAdd',
+            ['id_product_old' => $oldProductId, 'id_product' => $newProductId, 'product' => $newProduct]
+        );
+
+        $this->updateSearchIndexation($newProduct, $oldProductId);
+
+        return new ProductId((int) $newProduct->id);
+    }
+
+    /**
+     * @param Product $newProduct
+     * @param int $oldProductId
+     *
+     * @throws CannotUpdateProductException
+     */
+    private function updateSearchIndexation(Product $newProduct, int $oldProductId): void
+    {
+        if (!in_array($newProduct->visibility, ['both', 'search']) || !$this->isSearchIndexationOn) {
+            return;
+        }
+
+        try {
+            if (!Search::indexation(false, $newProduct->id)) {
+                throw new CannotUpdateProductException(
+                    sprintf('Cannot update search indexation when duplicating product %d', $oldProductId),
+                    CannotUpdateProductException::FAILED_UPDATE_SEARCH_INDEXATION
+                );
+            }
+        } catch (PrestaShopException $e) {
+            throw new CoreException(
+                sprintf('Error occurred when duplicating product %d. Failed to update search indexation', $oldProductId),
+                0,
+                $e
+            );
+        }
+    }
+
+    /**
+     * @param Product $product
+     *
+     * @return Product the new product
+     */
+    private function duplicateProduct(Product $product): Product
+    {
+        $this->setName($product);
+        $this->setPriceByShops($product);
+
+        unset($product->id, $product->id_product);
+        $product->indexed = false;
+        $product->active = false;
+
+        $this->productRepository->add($product)->getValue();
+
+        return $product;
+    }
+
+    /**
+     * Sets new name for duplicated product by modifying the existing name with a new pattern
+     *
+     * @param Product $product
+     */
+    private function setName(Product $product): void
+    {
+        //@todo: would it be possible to transle it or is it ok that each lang will contain this generic pattern?
+        //@todo: make the pattern modifiable (add configuration) ?
```

Comment:
If it's being translated then it's a translation key so we need to keep always the same